api/rtx/src/uvisor_allocator.c

```diff
@@ -0,0 +1,215 @@
+    size_t page_count = 0;
+
+    /* Calcutate the required total size considering page size and overhead */
+    for (int request = size; request > 0; request -= UVISOR_PAGE_SIZE, page_count++)
+    {
+        /* Adjust the requested and total sizes */
+        request += block_overhead;
+        size += block_overhead;
+    }
+    DPRINTF("uvisor_allocator_create_with_pages: Requesting %u pages for at least %uB\n", page_count, size);
+
+    /* Compute the maximum allocation within our blocks */
+    size_t maximum_allocation_size = UVISOR_PAGE_SIZE - block_overhead;
+    /* If the required maximum allocation is larger than we can provide, abort */
+    if (maximum_malloc_size > maximum_allocation_size) {
+        DPRINTF("uvisor_allocator_create_with_pages: Maximum allocation request %uB is larger then available %uB\n\n", maximum_malloc_size, maximum_allocation_size);
+        return NULL;
+    }
+
+    /* Compute the require memory size for the page table */
+    size_t allocator_type_size = sizeof(UvisorAllocatorInternal);
+    /* Add size for each additional page */
+    allocator_type_size += (page_count - 1) * sizeof(void *);
+    /* Allocate this much memory */
+    UvisorAllocatorInternal * const allocator = malloc(allocator_type_size);
+    /* If malloc failed, abort */
+    if (allocator == NULL) {
+        DPRINTF("uvisor_allocator_create_with_pages: allocator_t failed to be allocated!\n\n");
+        return NULL;
+    }
+
+    /* Prepare the page table */
+    allocator->table.page_size = UVISOR_PAGE_SIZE;
+    allocator->table.page_count = page_count;
+    /* Get me some pages */
+    if (uvisor_page_malloc((UvisorPageTable*)&(allocator->table))) {
+        DPRINTF("uvisor_allocator_create_with_pages: Not enough free pages available!\n\n");
+        free(allocator);
+        return NULL;
+    }
+
+    /* Initialize a MEMP structure in all pages */
+    for(size_t ii = 0; ii < page_count; ii++)
+    {
+        /* Add each page as a pool */
+        rt_init_mem(allocator->table.page_origins[ii], UVISOR_PAGE_SIZE);
+        DPRINTF("uvisor_allocator_create_with_pages: Created MEMP allocator %p with offset %d\n", allocator->table.page_origins[ii], 0);
+    }
+    DPRINTF("\n");
+    /* Aaaand across the line */
+    return (UvisorAllocator) allocator;
+}
+
+int uvisor_allocator_destroy(
+    UvisorAllocator allocator)
+{
+    DPRINTF("uvisor_allocator_destroy: Destroying MEMP allocator at %p\n", table(allocator)->page_origins[0]);
+
+    /* Check if we are working on statically allocated memory */
+    UvisorAllocatorInternal * alloc = (UvisorAllocatorInternal * const) allocator;
+    if (alloc->table.page_count == 0) {
+        DPRINTF("uvisor_allocator_destroy: %p is not page-backed memory, not freeing!\n", allocator);
+        return -1;
+    }
+
+    /* Free all pages */
+    if (uvisor_page_free(&(alloc->table))) {
+        DPRINTF("uvisor_allocator_destroy: Unable to free pages!\n\n");
+        return -1;
```

Comment:
Why not pass on the error returned from `uvisor_page_free` instead of replacing it with `-1`?